src/libutils/oglapphelpers/metalapp.mm

```diff
@@ -0,0 +1,391 @@
+            separator = ", ";
+        }
+        
+        for(unsigned int i = 0; i < shaderDesc->getNumTextures(); ++i)
+        {
+            const char* textureName;
+            const char* samplerName;
+            unsigned int width, height;
+            GpuShaderCreator::TextureType channel;
+            Interpolation interpolation;
+            
+            shaderDesc->getTexture(i, textureName, samplerName, width, height, channel, interpolation);
+            
+            if(height > 1)
+                main << ",    texture2d<float> ";
+            else
+                main << ",    texture1d<float> ";
+            main << textureName
+                 << " [[ texture("
+                 << std::to_string(tex_slot)
+                 << ") ]]\n";
+            
+            main << ",    sampler "
+                 << samplerName
+                 << " [[ sampler("
+                 << std::to_string(tex_slot++)
+                 << ") ]]\n";
+            
+            params << separator
+                   << textureName
+                   << ", "
+                   << samplerName;
+            
+            separator = ", ";
+        }
+        
+        if(uniformParams.str().size() > 0 || params.str().size() > 0)
+            separator = ", ";
+        
+        main << ")"
+                "{\n"
+                "    constexpr sampler s = sampler(address::clamp_to_edge);\n"
+                "    float4 inPixel = colorIn.sample(s, float2(in.texCoord0.x, in.texCoord0.y));\n"
+                "    return "
+             <<  shaderDesc->getFunctionName() << "(" << params.str() << uniformParams.str() << separator << "inPixel);\n"
+             << "}\n";
+        
+        main << std::endl;
+    }
+    else
+    {
+        throw Exception("Metal renderer can only consume MSL shaders");
+    }
+    
+    if(printShader())
+    {
+        std::cout << std::endl;
+        std::cout << "GPU Shader Program:" << std::endl;
+        std::cout << std::endl;
+        std::cout << main.str() << std::endl;
+        std::cout << std::endl;
+    }
+    
+    // Build the fragment shader program.
+    if(m_metalBuilder->buildPipelineStateObject(main.str().c_str()))
+    {
+        m_metalBuilder->applyColorCorrection(m_image->getMetalTextureHandle(),
+                                             m_outputImage->getMetalTextureHandle(),
+                                             m_outputImage->getWidth(),
+                                             m_outputImage->getHeight());
+    }
+}
+
+void MetalApp::redisplay()
+{
+    if(m_metalBuilder)
+    {
+        m_metalBuilder->applyColorCorrection(m_image->getMetalTextureHandle(),
+                                            m_outputImage->getMetalTextureHandle(),
+                                            m_outputImage->getWidth(),
+                                            m_outputImage->getHeight());
+    }
+    
+    if(!m_glStateBound)
+    {
+        prepareAndBindOpenGLState();
```

Comment:
Should the bindings be done before the `applyColorCorrection()` ?

Reply:
It actually doesn't matter. This sets OpenGL state and acts independent from applyColorCorrection.